File: Go/1884-EggDropWith2EggsAndNFloors/Solution.go
package lc1884

// Returns the minimum number of moves to determine with certainty the critical floor f_c,
// given two eggs.
//
// An egg dropped above any floor higher than the critical floor f_c will break, while an egg
// dropped on or below the critical floor f_c will not break and can be reused.
//
// Suppose the minimum number of moves to determine f_c given two eggs is x.
// Then, the claim is that the first floor to drop the egg from cannot exceed x.
//
// Proof (by contradiction):
//  1. Suppose for the sake of contradiction that the first floor picked exceeds x. That is
//     pick x + 1.
//  2. Then, there are two possible cases:
//     2.1.   Case 1: Egg breaks
//     2.1.1. Then, for the critical floor to be found with certainty, the remaining egg has to be dropped
//     sequentially from floor 1, till floor x.
//     2.1.2. This linear search takes x steps. Combined with the 1 step from dropping the first egg at floor x + 1,
//     the total number of moves to find the critical floor is x + 1, which contradicts the claim.
//  3. Thus, given two eggs, the claim that the first floor picked cannot exceed x holds true.
//
// Generalizing it to k number of eggs, the claim is that there exists an Optimal substructure.
//
// Optimal Substructure:
// Given k eggs and n floors:
// Suppose an egg is dropped at floor x, the minimum number of moves to determine f_c
// is the minimum of the maximum (certainty i.e. worst case) between:
//  1. the minimum number of moves to deterimne f_c with (k - 1) eggs and (x - 1) floors, and (egg breaks)
//  2. the minimum number of moves to determine f_c with k eggs and (n - x) floors. (egg does not break)
//
// Suppose the minimum number of moves to determine f_c is x. Then
func twoEggDrop(n int) int {
	MAX_FLOORS := 10000
	// Initialize DP array.
	dp := make([][]int, 2)
	dp[0] = make([]int, n+1)
	dp[1] = make([]int, n+1)

	// Initialize base cases
	// 1 egg, i floors => i steps
	for i := 0; i < n+1; i++ {
		dp[0][i] = i
		// dp[1][i] = 1000000
	}
	dp[1][1] = 1 // 2 eggs, 1 floor => 1 trial/step

	// For each i number of floors
	for i := 2; i < n+1; i++ {
		dp[1][i] = MAX_FLOORS
		// try dropping at x = {1, ..., i}
		for x := 1; x < i+1; x++ {
			// dp[1][x] = min(dp[1][x], 1+max(dp[0][x-1], dp[1][i-x]))
			dp[1][i] = min(dp[1][i], 1+max(dp[0][x-1], dp[1][i-x]))
		}
	}
	return dp[1][n]
}

File: C++/1884-EggDropWith2EggsAndNFloors/Solution.cpp
#include <algorithm>
#include <cstddef>
#include <vector>

class Solution {
 public:
  int twoEggDrop(int n) {
    // There exists a floor f such that any eggs dropped at a floor higher than
    // f will break, and any egg dropped at or below floor f will not break.
    //
    // From phrasing alone, seems like Binary Search is involved.
    // From experience, its DP!
    // For 2 eggs, the subproblems are straightforward.
    // Naively, drop egg at floor k.
    // 1. If break, then there are k - 1 floors to search (k-1 moves) with
    //    1 egg.
    // 2. No break, then k floors can be eliminated => n - k floors to search
    //    with 2 eggs.
    // Therefore, dp[i][j] represents the minimum number of moves to
    // determine with certainty the critical floor f with i eggs remaining
    // and j floors left.

    const int INF = 1000 + 1;

    // 2 eggs, n floors
    // Instead of initializing to INF, which would affect the subsequent
    // loop using std::max, set to 0 first.
    std::vector<std::vector<int>> dp(3, std::vector<int>(n + 1, 0));

    // base case
    for (int f = 0; f <= n; ++f) {
      // 1 egg, f floors means a linear search is necessary
      dp[1][f] = f;
    }

    // for each number of eggs
    for (int i = 2; i <= 2; ++i) {
      // for each number of floors
      for (int j = 1; j <= n; ++j) {
        dp[i][j] = INF;
        // Try dropping egg at each floor. However, instead of linearly building
        // the table from bottom up naively, notice that there is somewhat of
        // a monotonic property here. That is, as the floor that the egg is
        // dropped at increases:
        // 1. The number of moves required is non-decreasing if the egg breaks,
        //    as denoted by dp[i - 1][k - 1], where searching k - 1 floors with
        //    1 less egg is at least searching k floors + 1 move.
        // 2. The number of moves required is non-increasing if the egg does
        //    not break. Denoted by dp[i][j - k], where searching less floors
        //    with the same number of eggs allow for an at least
        //    better-than-linear number of moves.
        //
        // The key is to find the number of floors where these two monotonic
        // functions intersect, since we want the most minimum number of moves
        // to determine with CERTAINTY, i.e., the worse case.
        // I.e., think of it as finding the valley.
        //
        // Therefore, we can employ binary search to find valleys!
        int left = 1;
        int right = j;
        while (left < right) {
          int k = left + (right - left) / 2;

          int broken = dp[i - 1][k - 1];
          int notBroken = dp[i][j - k];

          // visualize the shape of the graphs
          if (broken < notBroken) {
            // means the k-th floor lie on the left of the valley.
            left = k + 1;
          } else {
            right = k;
          }
        }

        int worstCase = 1 + std::max(dp[i - 1][left - 1], dp[i][j - left]);
        dp[i][j] = std::min(dp[i][j], worstCase);
      }
    }
    return dp[2][n];
  }
};
